compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,608 @@
+                       newExprs, rewriter.getContext());
+    newIndexingMaps[inputIndex] = transposedMap;
+    genericOp.setIndexingMapsAttr(
+        rewriter.getAffineMapArrayAttr(newIndexingMaps));
+
+    genericOp.setOperand(inputIndex, transposeOp.getInput());
+    rewriter.finalizeRootUpdate(genericOp);
+    return success();
+  }
+};
+
+// Sinks a transpose to the input of a linalg named op. The conditions for the
+// rewrite are
+//   1) One of the input producers to the named op is a linalg.transpose
+//   2) The named op is generalizable (and is not a transpose)
+// The easiest way to get the rewrite we want then is to just try to generalize
+// all transposed named ops and let the generic pattern handle the actual
+// rewrite.
+class GeneralizeInputTransposedNamedOp
+    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
+public:
+  using OpInterfaceRewritePattern<linalg::LinalgOp>::OpInterfaceRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::LinalgOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(linalgOp)) {
+      return failure();
+    }
+    // Don't generalize transposes.
+    if (isa<linalg::TransposeOp>(linalgOp)) {
+      return failure();
+    }
+    bool hasTranspose = false;
+    for (Value input : linalgOp.getDpsInputs()) {
+      if (input.getDefiningOp<linalg::TransposeOp>()) {
+        hasTranspose = true;
+        break;
+      }
+    }
+    if (!hasTranspose) {
+      return failure();
+    }
+    if (failed(linalg::generalizeNamedOp(rewriter, linalgOp))) {
+      return failure();
+    }
+    return success();
+  }
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Linalg Named Op -> Named Op Conversions
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+template <typename OpTy, typename ReplTy, int64_t inputIdx, int64_t N>
+class NamedOpConversion : public OpRewritePattern<OpTy> {
+public:
+  using OpRewritePattern<OpTy>::OpRewritePattern;
+  NamedOpConversion(MLIRContext *ctx, std::array<int64_t, N> perm,
+                    PatternBenefit b = 1)
+      : OpRewritePattern<OpTy>(ctx, b), permutation(perm) {}
+
+  LogicalResult matchAndRewrite(OpTy namedOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(namedOp)) {
+      return failure();
+    }
+    Value input = namedOp.getInputs()[inputIdx];
+    SmallVector<NamedAttribute> attrs = getPrunedAttributeList(namedOp);
+    if (auto transpose = input.getDefiningOp<linalg::TransposeOp>()) {
+      if (llvm::all_of(llvm::zip_equal(transpose.getPermutation(), permutation),
+                       [](std::tuple<int64_t, int64_t> it) {
+                         return std::get<0>(it) == std::get<1>(it);
+                       })) {
```

Comment:
We can make it easier if the type of `permutation` is `SmallVector<int64_t>`. Then we can do something like:

```cpp
SmallVector<int64_t> transPerm(transpose.getPermutation());
if (transPerm == permutation) {
  ...
}
```

The other comment is that we prefer early-return style. This will save us level of nesting. E.g.,

```cpp
auto transpose = input.getDefiningOp<linalg::TransposeOp>();
if (!transpose) return failure();

SmallVector<int64_t> transPerm(transpose.getPermutation());
if (transPerm != permutation) return failure();

...
```